src/Features/Core/Portable/CodeStyle2/AbstractCodeStyle2Provider.Analysis.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.CodeStyle2
+{
+    internal abstract partial class AbstractCodeStyle2Provider<
+        TOptionKind, 
+        TSyntaxKind,
+        TCodeStyleProvider>
+        where TSyntaxKind : struct
+        where TCodeStyleProvider : AbstractCodeStyle2Provider<TOptionKind, TSyntaxKind, TCodeStyleProvider>, new()
+    {
+        private void DiagnosticAnalyzerInitialize(AnalysisContext context)
+        {
+            var analyzeCodeBlock = GetCodeBlockAction();
+            if (analyzeCodeBlock != null)
+            {
+                context.RegisterCodeBlockAction(c => AnalyzeIfEnabled(c, analyzeCodeBlock));
+            }
+
+            var (syntaxKinds, analyzeSyntax) = GetSyntaxNodeAction();
+            if (analyzeSyntax != null)
+            {
+                context.RegisterSyntaxNodeAction(c => AnalyzeIfEnabled(c, analyzeSyntax), syntaxKinds);
+            }
+
+            // check for all the other things that can be registered as well.
+        }
+
+        private void AnalyzeIfEnabled(CodeBlockAnalysisContext context, Action<CodeBlockAnalysisContext, TOptionKind> analyze)
+            => AnalyzeIfEnabled(context, analyze, context.Options, context.SemanticModel.SyntaxTree, context.CancellationToken);
+
+        private void AnalyzeIfEnabled(SyntaxNodeAnalysisContext context, Action<SyntaxNodeAnalysisContext, TOptionKind> analyze)
+            => AnalyzeIfEnabled(context, analyze, context.Options, context.SemanticModel.SyntaxTree, context.CancellationToken);
+
+        private void AnalyzeIfEnabled<TContext>(
+            TContext context, Action<TContext, TOptionKind> analyze,
+            AnalyzerOptions options, SyntaxTree syntaxTree, CancellationToken cancellationToken)
+        {
+            var optionSet = options.GetDocumentOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult();
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var optionValue = optionSet.GetOption(_option);
+            switch (optionValue.Notification.Severity)
+            {
+                case ReportDiagnostic.Error:
+                case ReportDiagnostic.Warn:
+                case ReportDiagnostic.Info:
+                    break;
+                default:
+                    // don't analyze if it's any other value.
+                    return;
+            }
+
+            analyze(context, optionValue.Value);
+        }
+
+        public abstract class DiagnosticAnalyzer : AbstractCodeStyleDiagnosticAnalyzer
```

Comment:
here are the impls of hte actual DiagnosticAnalyzer/CodeRefactoringProvider/CodeFixProvider. They just defer everything to the AbstractCodeStyleProvider. Ideally they would be sealed types. However, because people have to export DiagnosticAnalyzer/CodeRefactoringProvider/CodeFixProvider themselves, these are exposed and abstract solely for the purpose of creating an exported subclass. You'll see that at the bottom of this file.